File: web/src/course/trainerCatalog.ts
import type { CourseBundle, Exercise } from './types';

export type TrainerDomain = 'grammar' | 'reading' | 'writing';

interface TrainerTopicSpec {
  id: string;
  domain: 'grammar';
  level: string;
  title: string;
  summary: string;
  roadmapItemId: string;
  skillIds: string[];
  supplementalExercises: Exercise[];
}

interface TrainerCatalog {
  version: number;
  topics: TrainerTopicSpec[];
}

export interface TrainerTopic {
  id: string;
  domain: TrainerDomain;
  level: string;
  title: string;
  summary: string;
  roadmapItemId: string;
  exercises: Exercise[];
}

const WRITING_TYPES = new Set<Exercise['type']>([
  'fill_blank',
  'sentence_builder',
  'letter_unscramble',
]);

let catalogPromise: Promise<TrainerCatalog> | null = null;

export function loadTrainerCatalog(): Promise<TrainerCatalog> {
  if (catalogPromise) return catalogPromise;
  catalogPromise = fetch('/course/trainer_catalog.json')
    .then((response) => {
      if (!response.ok) throw new Error('Не удалось загрузить каталог Тренажёрки.');
      return response.json() as Promise<TrainerCatalog>;
    })
    .then((catalog) => {
      if (catalog.version !== 1 || !Array.isArray(catalog.topics)) {
        throw new Error('Каталог Тренажёрки повреждён.');
      }
      return catalog;
    })
    .catch((error) => {
      catalogPromise = null;
      throw error;
    });
  return catalogPromise;
}

/**
 * Собирает три режима из одного источника. Грамматика получает все упражнения
 * связанного skill и короткую проверку понятия. Čitanje оставляет тексты с
 * вопросами, Pisanje — задания, где ответ действительно нужно набрать или
 * собрать, а не просто выбрать.
 */
export function buildTrainerTopics(
  bundle: CourseBundle,
  catalog: TrainerCatalog,
): TrainerTopic[] {
  const bySkill = new Map<string, Exercise[]>();
  for (const unit of bundle.units) {
    for (const skill of unit.skills) {
      bySkill.set(
        skill.id,
        skill.lessons.flatMap((lesson) => lesson.exercises),
      );
    }
  }

  const result: TrainerTopic[] = [];
  for (const spec of catalog.topics) {
    const courseExercises = uniqueExercises(
      spec.skillIds.flatMap((skillId) => bySkill.get(skillId) ?? []),
    );
    const grammar = uniqueExercises([
      ...spec.supplementalExercises,
      ...courseExercises,
    ]);
    if (grammar.length > 0) {
      result.push({
        id: spec.id,
        domain: 'grammar',
        level: spec.level,
        title: spec.title,
        summary: spec.summary,
        roadmapItemId: spec.roadmapItemId,
        exercises: grammar,
      });
    }

    const reading = courseExercises.filter((exercise) => exercise.type === 'reading_qa');
    if (reading.length > 0) {
      result.push({
        id: `${spec.id}-reading`,
        domain: 'reading',
        level: spec.level,
        title: spec.title,
        summary: 'Текст на сербском с вопросами на понимание.',
        roadmapItemId: '',
        exercises: reading,
      });
    }

    const writing = courseExercises.filter((exercise) => WRITING_TYPES.has(exercise.type));
    if (writing.length > 0) {
      result.push({
        id: `${spec.id}-writing`,
        domain: 'writing',
        level: spec.level,
        title: spec.title,
        summary: 'Наберите форму или соберите сербское предложение.',
        roadmapItemId: '',
        exercises: writing,
      });
    }
  }
  return result;
}

function uniqueExercises(exercises: Exercise[]): Exercise[] {
  const seen = new Set<string>();
  return exercises.filter((exercise) => {
    if (seen.has(exercise.id)) return false;
    seen.add(exercise.id);
    return true;
  });
}

File: frontend/lib/course/models/trainer_catalog.dart
import 'dart:convert';

import 'package:flutter/services.dart';

import 'course.dart';
import 'exercise.dart';

enum TrainerDomain { grammar, reading, writing }

extension TrainerDomainLabel on TrainerDomain {
  String get label => switch (this) {
        TrainerDomain.grammar => 'Gramatika',
        TrainerDomain.reading => 'Čitanje',
        TrainerDomain.writing => 'Pisanje',
      };
}

class TrainerTopicSpec {
  const TrainerTopicSpec({
    required this.id,
    required this.domain,
    required this.level,
    required this.title,
    required this.summary,
    required this.roadmapItemId,
    required this.skillIds,
    required this.supplementalExercises,
  });

  final String id;
  final TrainerDomain domain;
  final String level;
  final String title;
  final String summary;
  final String roadmapItemId;
  final List<String> skillIds;
  final List<Exercise> supplementalExercises;

  factory TrainerTopicSpec.fromJson(Map<String, dynamic> json) =>
      TrainerTopicSpec(
        id: (json['id'] ?? '').toString(),
        domain: switch ((json['domain'] ?? '').toString()) {
          'grammar' => TrainerDomain.grammar,
          'reading' => TrainerDomain.reading,
          'writing' => TrainerDomain.writing,
          final value =>
            throw FormatException('Неизвестный раздел Тренажёрки: $value'),
        },
        level: (json['level'] ?? '').toString(),
        title: (json['title'] ?? '').toString(),
        summary: (json['summary'] ?? '').toString(),
        roadmapItemId: (json['roadmapItemId'] ?? '').toString(),
        skillIds: (json['skillIds'] as List? ?? const [])
            .map((item) => item.toString())
            .toList(),
        supplementalExercises:
            (json['supplementalExercises'] as List? ?? const [])
                .whereType<Map>()
                .map((item) => Exercise.fromJson(item.cast<String, dynamic>()))
                .toList(),
      );
}

class TrainerTopic {
  const TrainerTopic({
    required this.id,
    required this.domain,
    required this.level,
    required this.title,
    required this.summary,
    required this.roadmapItemId,
    required this.exercises,
  });

  final String id;
  final TrainerDomain domain;
  final String level;
  final String title;
  final String summary;
  final String roadmapItemId;
  final List<Exercise> exercises;
}

Future<List<TrainerTopicSpec>> loadTrainerCatalog() async {
  final raw = await rootBundle.loadString('assets/course/trainer_catalog.json',
      cache: true);
  final json = (jsonDecode(raw) as Map).cast<String, dynamic>();
  if (json['version'] != 1) {
    throw const FormatException('Неизвестная версия каталога Тренажёрки.');
  }
  return (json['topics'] as List? ?? const [])
      .whereType<Map>()
      .map((item) => TrainerTopicSpec.fromJson(item.cast<String, dynamic>()))
      .toList();
}

List<TrainerTopic> buildTrainerTopics(
  Course course,
  List<TrainerTopicSpec> catalog,
) {
  final bySkill = <String, List<Exercise>>{};
  for (final unit in course.units) {
    for (final skill in unit.skills) {
      bySkill[skill.id] = [
        for (final lesson in skill.lessons) ...lesson.exercises,
      ];
    }
  }

  final result = <TrainerTopic>[];
  for (final spec in catalog) {
    final courseExercises = _unique([
      for (final skillId in spec.skillIds) ...bySkill[skillId] ?? const [],
    ]);
    final exercises = _unique([
      ...spec.supplementalExercises,
      ...courseExercises,
    ]);
    if (exercises.isNotEmpty) {
      result.add(TrainerTopic(
        id: spec.id,
        domain: spec.domain,
        level: spec.level,
        title: spec.title,
        summary: spec.summary,
        roadmapItemId: spec.roadmapItemId,
        exercises: exercises,
      ));
    }
  }
  return result;
}

List<Exercise> _unique(Iterable<Exercise> exercises) {
  final seen = <String>{};
  return [
    for (final item in exercises)
      if (seen.add(item.id)) item
  ];
}
